cpp/include/cudf/io/experimental/hybrid_scan.hpp

```diff
@@ -0,0 +1,253 @@
+namespace io::parquet::experimental::detail {
+/**
+ * @brief Internal experimental Parquet reader optimized for highly selective filters, called a
+ *        Hybrid Scan operation.
+ */
+class hybrid_scan_reader_impl;
+}  // namespace io::parquet::experimental::detail
+}  // namespace CUDF_EXPORT cudf
+
+//! Using `byte_range_info` from cudf::io::text
+using cudf::io::text::byte_range_info;
+
+namespace CUDF_EXPORT cudf {
+namespace io::parquet::experimental {
+/**
+ * @addtogroup io_readers
+ * @{
+ * @file
+ */
+
+/**
+ * @brief The experimental parquet reader class to optimally read parquet files subject to
+ *        highly selective filters, called a Hybrid Scan operation
+ *
+ * This class is designed to best exploit reductive optimization techniques to speed up reading
+ * Parquet files subject to highly selective filters. This class reads file contents in two passes,
+ * where the first pass optimally reads the `filter` columns (i.e. columns that appear in the filter
+ * expression) and the second pass optimally reads the `payload` columns (i.e. columns that do not
+ * appear in the filter expression).
+ */
+class hybrid_scan_reader {
+ public:
+  /**
+   * @brief Constructor for the experimental parquet reader class to optimally read Parquet files
+   * subject to highly selective filters
+   *
+   * @param footer_bytes Host span of parquet file footer bytes
+   * @param options Parquet reader options
+   */
+  explicit hybrid_scan_reader(cudf::host_span<uint8_t const> footer_bytes,
+                              parquet_reader_options const& options);
+
+  /**
+   * @brief Destructor for the experimental parquet reader class
+   */
+  ~hybrid_scan_reader();
+
+  /**
+   * @brief Get the Parquet file footer metadata
+   *
+   * Returns the materialized Parquet file footer metadata struct. The footer will contain the
+   * materialized page index if called after `setup_page_index()`.
+   *
+   * @return Parquet file footer metadata
+   */
+  [[nodiscard]] FileMetaData const& get_parquet_metadata() const;
+
+  /**
+   * @brief Get the byte range of the page index in the Parquet file
+   *
+   * @return Byte range of the page index
+   */
+  [[nodiscard]] byte_range_info get_page_index_bytes() const;
+
+  /**
+   * @brief Setup the page index within the Parquet file metadata (`FileMetaData`) for later use
+   *
+   * Materialize the `ColumnIndex` and `OffsetIndex` structs (collectively called the page index)
+   * within the Parquet file metadata struct. The statistics contained in page index can be used to
+   * prune data pages before decoding
+   *
+   * @param page_index_bytes Host span of Parquet page index buffer bytes
+   */
+  void setup_page_index(cudf::host_span<uint8_t const> page_index_bytes) const;
+
+  /**
+   * @brief Get all available row groups from the parquet file
+   *
+   * @param options Parquet reader options
+   * @return Vector of row group indices
+   */
+  [[nodiscard]] std::vector<size_type> get_all_row_groups(
```

Comment:
can this just return the number of row groups? is the returned vector ever different from `iota`?
Edit: ah, is it so that a subspan of the returned vector can be passed to filter functions?

Reply:
Yeah exactly. Just for the semantics. You can just pass the output from here to one of the filter APIs without intermediate stuff.